src/main/scala/vectorpipe/vectortile/Pipeline.scala

```diff
@@ -0,0 +1,126 @@
+package vectorpipe.vectortile
+
+import geotrellis.spark.SpatialKey
+import geotrellis.spark.tiling._
+import geotrellis.vector.{Feature, Geometry}
+import geotrellis.vectortile.Value
+import org.apache.spark.sql.{DataFrame, Row}
+import org.locationtech.jts.{geom => jts}
+import org.locationtech.jts.simplify
+
+/**
+ * The interface governing the transformation from processed OSM dataframes to
+ * vector tiles.
+ *
+ * After loading an OSM source and transforming via [[OSM.toGeometry]] and any
+ * filtering/modification steps, one may want to output a set of vector tiles
+ * from the remaining geometry and metadata.  The vector tile output process is
+ * governed by a Pipeline definition that describes how to convert a data frame
+ * into a set of vector tiles and how to pass information to subsequent zoom
+ * levels.
+ *
+ * [It is worth noting that nothing in this interface restricts the input data
+ * to having originated from OSM.  Any DataFrame containing a column of JTS
+ * geometry is a valid target.]
+ *
+ * This API treats vector tile generation as an inherently hierarchical process
+ * governed by a series of operations.  Every zoom level will be addressed in
+ * the same way, and the sequence of operations can be depicted schematically as
+ * below:
+ *
+ *     ┌──────────┐
+ *     │ Zoom z-1 │
+ *     └──────────┘
+ *          ↑
+ *       simplify
+ *          ↑
+ *        reduce
+ *          ↑
+ *     ╔══════════╗                          ┌─────────────┐
+ *     ║  Zoom z  ║ → select → clip → pack → │ VECTOR TILE │
+ *     ╚══════════╝                          └─────────────┘
+ *
+ * Each of these operations is documented below.  The only extra note is that [[
+ * reduce]] and [[simplify]] will be called prior to processing the initial zoom
+ * level.
+ */
+trait Pipeline {
+  /**
+   * The root URI for output.
+   */
+  val baseOutputURI: java.net.URI
+
+  /**
+   * Name of the column containing the JTS geometry.
+   */
+  val geometryColumn: String
+
+  /**
+   * Reduce the input data between zoom levels.
+   *
+   * Not all data is useful in all levels of the vector tile pyramid.  When a
+   * geometry will no longer be used in any subsequent generated level, it may be
+   * desirable to drop the relevant rows from the DataFrame to limit the amount
+   * of working data.  Moreover, it may be desirable to collect multiple entities
+   * into a single, aggregated entity.  These sorts of operations are made
+   * possible by the [[reduce]] function.
+   *
+   * This function will be called once before tile generation begins at the
+   * initial zoom level.  All participating geometries will already have been
+   * keyed to the given layout, with the list of relevant keys available in the
+   * "keys" field.
```

Comment:
Can you elaborate on what the "relevant keys" are? Are these spatial keys / tile coordinates, or are they primary keys relative to the dataset itself?